File: backend/read_excel.py
import pandas as pd
import json

df = pd.read_excel('../Kiểm kê 2024.xlsx', sheet_name='CL', header=0)
# Forward fill the "Đơn vị" column because of merged cells
df['Đơn vị'] = df['Đơn vị'].ffill()

# Drop rows where "Bộ Phận" is empty
df = df.dropna(subset=['Bộ Phận'])

unique_pairs = df[['Đơn vị', 'Bộ Phận']].drop_duplicates()

departments_map = {}
for index, row in unique_pairs.iterrows():
    unit = str(row['Đơn vị']).strip()
    dept = str(row['Bộ Phận']).strip()
    
    # Exclude headers themselves if they got caught
    if unit == 'Đơn vị' or dept == 'Bộ Phận':
        continue
        
    if pd.isna(row['Đơn vị']) or pd.isna(row['Bộ Phận']):
        continue
    
    if unit not in departments_map:
        departments_map[unit] = set()
    departments_map[unit].add(dept)

result = []
for unit, depts in departments_map.items():
    result.append({
        'group': unit,
        'items': sorted(list(depts))
    })

result = sorted(result, key=lambda x: x['group'])
print(json.dumps(result, ensure_ascii=False, indent=2))

File: backend/read_excel.js
import ExcelJS from 'exceljs';

const workbook = new ExcelJS.Workbook();
await workbook.xlsx.readFile('../Kiểm kê 2024.xlsx');
const sheet = workbook.getWorksheet('CL') || workbook.worksheets[0];
const data = [];

sheet.eachRow({ includeEmpty: true }, (row) => {
    data.push(row.values.slice(1));
});

const departmentsMap = {};
let currentUnit = null;

// Skip headers, assuming data starts at index 3
for (let i = 3; i < data.length; i++) {
    const row = data[i];
    if (!row) continue;
    
    const unit = row[0];
    const department = row[1];
    
    // Update currentUnit if it's not empty
    if (unit && typeof unit === 'string' && unit.trim() !== '') {
        currentUnit = unit.trim();
    }
    
    if (currentUnit && department && typeof department === 'string' && department.trim() !== '') {
        const cleanDept = department.trim();
        
        if (!departmentsMap[currentUnit]) {
            departmentsMap[currentUnit] = new Set();
        }
        departmentsMap[currentUnit].add(cleanDept);
    }
}

const result = [];
for (const [unit, depts] of Object.entries(departmentsMap)) {
    result.push({
        group: unit,
        items: Array.from(depts).sort((a, b) => a.localeCompare(b, 'vi'))
    });
}

result.sort((a, b) => a.group.localeCompare(b.group, 'vi'));

console.log(JSON.stringify(result, null, 2));
